### PR TITLE
Bucket ref protection on replicas - part 3 - consult replicas before making a bucket GARBAGE

### PR DESCRIPTION
**Pull request by Gerold103**
The patchset is a next step on the way of fixing the buckets being deleted on master while still having RO refs on replicas.

It makes the master talk to replicas about which buckets they allow to GC, i.e. which of them have no RO refs anymore. The consulting is done in batches, in background, and doesn't affect incoming user requests.

The next part will be about introducing stricter `_bucket` changes validation via `on_replace` triggers.

CI sometimes can fail due to https://github.com/tarantool/tarantool/issues/7487.

### CODE REVIEW

File: test/instances/storage.lua [hunk 1]
```diff
@@ -108,6 +108,18 @@ local function bucket_gc_continue()
     vshard.storage.garbage_collector_wakeup()
```
Comment (olegrok) [inline test/instances/storage.lua:108]: > This should be fixed separately. Most likely by addition of sanity
checks to the _bucket:on_replace trigger to raise an error when
there is a threat of data loss or corruption.

I suggest to file an issue just in order not to forget

**Review by Serpentian — COMMENTED** [review]
Thanks for the patch!

File: vshard/util.lua [hunk 2]
```diff
@@ -342,6 +342,7 @@ local feature = {
         return version_is_at_least(2, 10, 0, nil, 0, 2)
     end)(),
     error_stack = version_is_at_least(2, 4, 0, nil, 0, 0),
+    csw = fiber.self().csw ~= nil,
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply vshard/util.lua:345]: > I don't understand how it's related to patch

I need it in a new test. Didn't seem worth a separate commit.

> And probably this condition is always true.

It is `false` on 1.10.
```
Tarantool 1.10.13-18-g588c65383
type 'help' for interactive help
tarantool> fiber = require('fiber')
---
...

tarantool> fiber.self().csw
---
- null
...

tarantool> fiber.self().csw ~= nil
---
- false
...
```

> Maybe you want to check fiber.csw from 2.10+?

I just need any way to get `csw` except `fiber.info()`. Both `fiber_object:csw()` and `fiber_lib.csw()` are absent on 1.10.

File: test/instances/storage.lua (hunk 1 above)
Comment (Gerold103) [reply test/instances/storage.lua:108]: It is the same issue. I didn't close 173th by this PR. There will be part 4.

File: test/storage-luatest/garbage_collector_test.lua [hunk 3]
```diff
@@ -597,3 +597,132 @@ test_group.test_unit_bucket_delete_garbage = function(g)
...
+                end
+                _bucket:replace({i, ivconst.BUCKET.SENT})
+            end
+        end
+        box.commit()
+        if ivutil.feature.csw then
+            local csw1 = ifiber.self():csw()
+            ilt.assert_equals(test_gc(sent_bids), sent_ro_bids)
+            local csw2 = ifiber.self():csw()
+            ilt.assert_equals(csw2, csw1 + 2)
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply test/storage-luatest/garbage_collector_test.lua:947]: `_bucket` is always memtx. Cfg engine only affects `box.space.test` which is not used here.

File: vshard/util.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (olegrok) [reply vshard/util.lua:345]: Ok. Thanks for explanation.

File: test/instances/storage.lua (hunk 1 above)
Comment (olegrok) [reply test/instances/storage.lua:108]: Ok

**Review by olegrok — APPROVED** [review]

**Review by Serpentian — APPROVED** [review]

File: test/replicaset-luatest/replicaset_3_test.lua [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    --
+    -- Too long request.
+    --
+    vtest.cluster_exec_each(g, function(uuid_c)
+        rawset(_G, 'test_sleep_do', true)
+        rawset(_G, 'test_sleep_c', function()
+            rawset(_G, 'test_sleep_is_called', true)
+            if box.info.uuid == uuid_c then
+                while _G.test_sleep_do do
+                    ifiber.sleep(0.001)
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply test/replicaset-luatest/replicaset_3_test.lua]: You approved the PR too early, I didn't even manage to respond to your comments yet 😀.
1 millisecond busy loop won't occupy even 1% of a single core. More like 0.01% or less. But I don't mind to change it. I introduced a named constant for busy loops which is 5 milliseconds. Should be more than enough.
<details><summary>diff</summary>

```Diff
diff --git a/test/luatest_helpers/vtest.lua b/test/luatest_helpers/vtest.lua
index 943c202..2d8bab2 100644
--- a/test/luatest_helpers/vtest.lua
+++ b/test/luatest_helpers/vtest.lua
@@ -8,6 +8,8 @@ local yaml = require('yaml')
 local vrepset = require('vshard.replicaset')
 
 local wait_timeout = 120
+-- Use it in busy-loops like `while !cond do fiber.sleep(busy_step) end`.
+local busy_step = 0.005
 local uuid_idx = 1
 --
 -- The maps help to preserve the same UUID for replicas and replicasets during
@@ -612,4 +614,5 @@ return {
     router_disconnect = router_disconnect,
     uuid_from_int = uuid_from_int,
     wait_timeout = wait_timeout,
+    busy_step = busy_step,
 }
diff --git a/test/replicaset-luatest/replicaset_3_test.lua b/test/replicaset-luatest/replicaset_3_test.lua
index 95ffd81..6634a46 100644
--- a/test/replicaset-luatest/replicaset_3_test.lua
+++ b/test/replicaset-luatest/replicaset_3_test.lua
@@ -85,7 +85,7 @@ test_group.test_replicaset_map_call = function(g)
             rawset(_G, 'test_sleep_is_called', true)
             if box.info.uuid == uuid_c then
                 while _G.test_sleep_do do
-                    ifiber.sleep(0.001)
+                    ifiber.sleep(ivtest.busy_step)
                 end
             end
             return true
```
</details>

File: vshard/replicaset.lua [hunk 5]
```diff
@@ -604,6 +604,63 @@ local function replicaset_template_multicallro(prefer_replica, balance)
...
+    local _, res, err, err_uuid, map
+    local replica_count = 0
+    --
+    -- Send requests.
+    --
+    local futures = {}
+    local opts_call = {is_async = true, timeout = 0}
+    for uuid, replica in pairs(replicaset.replicas) do
+        local conn = replicaset_connect_to_replica(replicaset, replica)
+        timeout, err = netbox_wait_connected(conn, timeout)
```
[the remark this answers is not in the document]
Comment (Gerold103) [reply vshard/replicaset.lua]: I agree with Oleg here. I was just too lazy to do it in the beginning. Now I introduced `replicaset:wait_connected_all()` and used it in `replicaset:map_call()`. See the updated commits, it is spread across 2.

**Pull request merged**